### PR TITLE
Mosaic#tabs

### PR DESCRIPTION
**Pull request by denismaxim0v**
this is still a wip

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -194,8 +194,10 @@ pub fn start(mut os_input: Box<dyn OsApi>, opts: Opt) {
                 let mut command_is_executing = command_is_executing.clone();
                 move || {
                     if let Some(layout) = maybe_layout {
+                        pty_bus.spawn_tab();
                         pty_bus.spawn_terminals_for_layout(layout);
                     } else {
+                        pty_bus.spawn_tab();
```
Comment (kunalmohan) [inline src/main.rs:200]: I don't think we need this here. When we create a new `Screen` in the beginning of `start()` function, we create a new tab as well (inside `Screen::new()`). Spawning a new tab here would mean creating two tabs at startup.

Comment (denismaxim0v) [reply src/main.rs:200]: Hmm, I guess I forgot to clean that up, we aren't spawning a tab inside Screen::new() anymore actually

**Review by imsnif — CHANGES_REQUESTED** [review]
@denismaxim0v + @kunalmohan - this is great. The implementation is very thought through and the code looks really clean. I'm happy to see your joint work on this.

I left some comments in the code - nothing major.
Other than that, do you think we can add some tests for this feature specifically?
A few ideas for tests:
* Open two tabs, open some panes in one tab, switch between them
* Open three tabs, close the middle one, switch between them
* Open two tabs, open some panes in one of them, switch to full screen, switch to the second tab, switch back
* more? I guess you probably have some interesting corners you can think of testing

I think you can get inspiration from the current tests (I know @denismaxim0v wrote some good tests in the past), but let me know if you'd like a hand.

File: src/screen.rs [hunk 2]
```diff
@@ -1623,3 +1649,132 @@ impl Screen {
...
+        Screen {
+            receiver: receive_screen_instructions,
+            max_panes,
+            send_pty_instructions,
+            send_app_instructions,
+            full_screen_ws: *full_screen_ws,
+            active_tab: None,
+            tabs: BTreeMap::new(),
+            os_api,
+            next_tab_index: 0,
```
[the remark this answers is not in the document]
Comment (kunalmohan) [reply src/screen.rs:1671]: Oh right we should do that. It didn't occur to me at the time.

File: src/screen.rs [hunk 3]
```diff
@@ -75,55 +75,86 @@ pub enum ScreenInstruction {
     ToggleActiveTerminalFullscreen,
     ClosePane(RawFd),
     ApplyLayout((Layout, Vec<RawFd>)),
+    NewTab(RawFd),
+    SwitchTabNext,
+    SwitchTabPrev,
+    CloseTab,
+}
+
+pub struct Tab {
```
[the remark this answers is not in the document]
Comment (denismaxim0v) [reply src/screen.rs:84]: > What do you think about moving this to a different file? `tab.rs`?

Yes, don't see why not

**Comment by imsnif** [comment]
@denismaxim0v, @kunalmohan, can you let me know when this is ready for another review?

**Review by imsnif — APPROVED** [review]
Very very cool. Good job on the quick adjustments! Looking forward to seeing more features from you two (together or separately) :)

**Pull request merged**